runtime/interpreter/storagemap.go

```diff
@@ -0,0 +1,243 @@
+//
+func (s StorageMap) ReadValue(key string) OptionalValue {
+	storable, err := s.orderedMap.Get(
+		stringAtreeComparator,
+		stringAtreeHashInput,
+		stringAtreeValue(key),
+	)
+	if err != nil {
+		if _, ok := err.(*atree.KeyNotFoundError); ok {
+			return NilValue{}
+		}
+		panic(ExternalError{err})
+	}
+
+	value := StoredValue(storable, s.orderedMap.Storage)
+	return NewSomeValueNonCopying(value)
+}
+
+// WriteValue sets or removes a value in the storage map.
+// If the given value is a SomeValue, the key is updated.
+// If the given value is NilValue, the key is removed.
+//
+func (s StorageMap) WriteValue(interpreter *Interpreter, key string, value OptionalValue) {
+	switch value := value.(type) {
+	case *SomeValue:
+		s.setValue(interpreter, key, value.Value)
+	case NilValue:
+		s.removeValue(interpreter, key)
+	default:
+		panic(atree.NewUnreachableError())
+	}
+}
+
+// setValue sets a value in the storage map.
+// If the given key already stores a value, it is overwritten.
+//
+func (s StorageMap) setValue(interpreter *Interpreter, key string, value Value) {
+	existingStorable, err := s.orderedMap.Set(
+		stringAtreeComparator,
+		stringAtreeHashInput,
+		stringAtreeValue(key),
+		value,
+	)
+	if err != nil {
+		panic(ExternalError{err})
+	}
+	interpreter.maybeValidateAtreeValue(s.orderedMap)
+
+	if existingStorable != nil {
+		existingValue := StoredValue(existingStorable, interpreter.Storage)
+		existingValue.DeepRemove(interpreter)
+		interpreter.RemoveReferencedSlab(existingStorable)
+	}
+}
+
+// removeValue removes a value in the storage map, if it exists.
+//
+func (s StorageMap) removeValue(interpreter *Interpreter, key string) {
+	existingKeyStorable, existingValueStorable, err := s.orderedMap.Remove(
+		stringAtreeComparator,
+		stringAtreeHashInput,
+		stringAtreeValue(key),
+	)
+	if err != nil {
+		if _, ok := err.(*atree.KeyNotFoundError); ok {
+			return
+		}
+		panic(ExternalError{err})
+	}
+	interpreter.maybeValidateAtreeValue(s.orderedMap)
+
+	// Key
+
+	// NOTE: key / field name is stringAtreeValue,
+	// and not a Value, so no need to deep remove
+	interpreter.RemoveReferencedSlab(existingKeyStorable)
+
+	// Value
+
+	if existingValueStorable != nil {
+		existingValue := StoredValue(existingValueStorable, interpreter.Storage)
+		existingValue.DeepRemove(interpreter)
+		interpreter.RemoveReferencedSlab(existingValueStorable)
+	}
+}
+
+// Iterator returns an iterator (StorageMapIterator),
+// which allows iterating over the keys and values of the storage map
+//
+func (s StorageMap) Iterator() StorageMapIterator {
+	mapIterator, err := s.orderedMap.Iterator()
+	if err != nil {
+		panic(ExternalError{err})
+	}
+
+	return StorageMapIterator{
+		mapIterator: mapIterator,
+		storage:     s.orderedMap.Storage,
+	}
+}
+
+func (s StorageMap) StorageID() atree.StorageID {
+	return s.orderedMap.StorageID()
+}
+
+// StorageMapIterator is an iterator over StorageMap
+//
+type StorageMapIterator struct {
```

Comment:
This iterator will be used in the future for exposing an storage iteration API to Cadence programs